lib/spack/spack/modules.py

```diff
@@ -259,3 +259,126 @@ def _write(self, m_file):
                 m_file.write("prepend-path %s \"%s\"\n" % (var, directory))
 
         m_file.write("prepend-path CMAKE_PREFIX_PATH \"%s\"\n" % self.spec.prefix)
+
+
+import spack.compilers
+
+
+class LmodModule(EnvModule):
+    name = 'lmod'
+    path = join_path(spack.share_path, "lmod", "modulefiles")
+
+    def __init__(self, spec=None):
+        super(LmodModule, self).__init__(spec)
+        # Sets tha appropriate category to be used with the 'family' function
+        if self.spec.name in spack.compilers.supported_compilers():
+            self.family = 'compiler'
+        elif self.spec.package.provides('mpi'):
+            self.family = 'mpi'
+        else:
+            self.family = None
+
+    @property
+    def file_name(self):
+        modulefiles_name = LmodModule.path
+        if self._use_system_compiler():
+            # If the module is installed using the system compiler put the modulefile in 'Core'
+            hierarchy_name = 'Core'
+        elif not self._is_mpi_dependent():
+            # If the module is serial and built using a compiler other than the system one,
+            # put the modulefile in '<Compiler>/<Version>'
+            hierarchy_name = self._compiler_module_directory(self.spec.compiler.name, self.spec.compiler.version)
+        elif self._is_mpi_dependent() and not self._use_system_compiler():
+            # If the module is MPI parallel then put the modulefile in '<MPI>/<MPI-Version>/<Compiler/<Compiler-Version>'
+            cname = self.spec.compiler.name
+            cversion = self.spec.compiler.version
+            for key, val in self.spec.dependencies.items():
+                package = spack.db.get(str(val))
+                if package.provides('mpi'):
+                    mpi_name = package.name
+                    mpi_version = package.version
+                    break
+            else:
+                raise RuntimeError('No MPI dependency found')
```

Comment:
@alalazo: I think this will throw an exception if you install anything that doesn't depend on MPI, right?   I guess that needs polishing.